### ISSUE TITLE
Runner scale up failed and the error is not quite clear

### ISSUE THREAD

**Issue by chenrui333** [summerwind/actions-runner-controller#34]
Today, I am trying to scale up my runner fleet, the runner container just self-restarted like this

```
Events:
  Type     Reason     Age                   From                                   Message
  ----     ------     ----                  ----                                   -------
  Normal   Scheduled  7m29s                 default-scheduler                      Successfully assigned default/meetup-android-runner-f6zfn-hvbjm to ip-10-181-3-182.ec2.internal
  Normal   Pulled     7m28s                 kubelet, ip-10-181-3-182.ec2.internal  Container image "docker:19.03.6-dind" already present on machine
  Normal   Created    7m28s                 kubelet, ip-10-181-3-182.ec2.internal  Created container docker
  Normal   Started    7m27s                 kubelet, ip-10-181-3-182.ec2.internal  Started container docker
  Warning  BackOff    2m24s                 kubelet, ip-10-181-3-182.ec2.internal  Back-off restarting failed container
  Normal   Created    2m9s (x3 over 7m28s)  kubelet, ip-10-181-3-182.ec2.internal  Created container runner
  Normal   Started    2m9s (x3 over 7m28s)  kubelet, ip-10-181-3-182.ec2.internal  Started container runner
  Normal   Pulling    2m9s (x3 over 7m28s)  kubelet, ip-10-181-3-182.ec2.internal  Pulling image "chenrui/actions-runner:v2.168.0"
  Normal   Pulled     2m9s (x3 over 7m28s)  kubelet, ip-10-181-3-182.ec2.internal  Successfully pulled image "chenrui/actions-runner:v2.168.0"
```

When I tried to check with the logs, I could not find anything useful (with `runner`, it is just info logs, with `docker`, it has something like this):

```
time="2020-04-15T16:30:44.882673953Z" level=info msg="loading plugin "io.containerd.metadata.v1.bolt"..." type=io.containerd.metadata.v1
time="2020-04-15T16:30:44.882858479Z" level=warning msg="could not use snapshotter btrfs in metadata plugin" error="path /var/lib/docker/containerd/daemon/io.containerd.snapshotter.v1.btrfs must be a btrfs filesystem to be used with the btrfs snapshotter"
time="2020-04-15T16:30:44.883115566Z" level=warning msg="could not use snapshotter aufs in metadata plugin" error="modprobe aufs failed: "ip: can't find device 'aufs'\nmodprobe: can't change directory to '/lib/modules': No such file or directory\n": exit status 1"
time="2020-04-15T16:30:44.883295667Z" level=warning msg="could not use snapshotter zfs in metadata plugin" error="path /var/lib/docker/containerd/daemon/io.containerd.snapshotter.v1.zfs must be a zfs filesystem to be used with the zfs snapshotter: skip plugin"
...
time="2020-04-15T16:30:45.174440728Z" level=info msg="Loading containers: start."
time="2020-04-15T16:30:45.244227388Z" level=warning msg="Running modprobe bridge br_netfilter failed with message: ip: can't find device 'bridge'\nbridge                172032  1 br_netfilter\nstp                    16384  1 bridge\nllc                    16384  2 bridge,stp\nipv6                  524288 408 nf_conntrack_ipv6,nf_nat_ipv6,bridge,[permanent]\nip: can't find device 'br_netfilter'\nbr_netfilter           24576  0 \nbridge                172032  1 br_netfilter\nmodprobe: can't change directory to '/lib/modules': No such file or directory\n, error: exit status 1"
time="2020-04-15T16:30:45.540762014Z" level=info msg="Default bridge (docker0) is assigned with an IP address 172.17.0.0/16. Daemon option --bip can be used to set a preferred IP address"
```

**Comment by chenrui333** [comment]
Tested with the other runner, it seems working fine:

```
√ Connected to GitHub
2020-04-15 15:45:44Z: Runner reconnected.
2020-04-15 15:45:44Z: Listening for Jobs
2020-04-15 16:54:28Z: Running job: terraform
2020-04-15 16:55:20Z: Job terraform completed with result: Succeeded
```

**Comment by chenrui333** [comment]
here is the full log:
<details>
<summary>full log</summary>

```
Signature ok
subject=CN = docker:dind client
Getting CA Private Key
/certs/client/cert.pem: OK
time="2020-04-15T17:57:01.905239818Z" level=info msg="Starting up"
time="2020-04-15T17:57:01.907392150Z" level=warning msg="could not change group /var/run/docker.sock to docker: group docker not found"
time="2020-04-15T17:57:01.908958121Z" level=info msg="libcontainerd: started new containerd process" pid=52
time="2020-04-15T17:57:01.909028418Z" level=info msg="parsed scheme: \"unix\"" module=grpc
time="2020-04-15T17:57:01.909047549Z" level=info msg="scheme \"unix\" not registered, fallback to default scheme" module=grpc
time="2020-04-15T17:57:01.909082849Z" level=info msg="ccResolverWrapper: sending update to cc: {[{unix:///var/run/docker/containerd/containerd.sock 0  <nil>}] <nil>}" module=grpc
time="2020-04-15T17:57:01.909116959Z" level=info msg="ClientConn switching balancer to \"pick_first\"" module=grpc
time="2020-04-15T17:57:01.937120511Z" level=info msg="starting containerd" revision=35bd7a5f69c13e1563af8a93431411cd9ecf5021 version=v1.2.12
time="2020-04-15T17:57:01.937694134Z" level=info msg="loading plugin "io.containerd.content.v1.content"..." type=io.containerd.content.v1
time="2020-04-15T17:57:01.937872799Z" level=info msg="loading plugin "io.containerd.snapshotter.v1.btrfs"..." type=io.containerd.snapshotter.v1
time="2020-04-15T17:57:01.938178407Z" level=warning msg="failed to load plugin io.containerd.snapshotter.v1.btrfs" error="path /var/lib/docker/containerd/daemon/io.containerd.snapshotter.v1.btrfs must be a btrfs filesystem to be used with the btrfs snapshotter"
time="2020-04-15T17:57:01.938207120Z" level=info msg="loading plugin "io.containerd.snapshotter.v1.aufs"..." type=io.containerd.snapshotter.v1
time="2020-04-15T17:57:01.949508715Z" level=warning msg="failed to load plugin io.containerd.snapshotter.v1.aufs" error="modprobe aufs failed: "ip: can't find device 'aufs'\nmodprobe: can't change directory to '/lib/modules': No such file or directory\n": exit status 1"
time="2020-04-15T17:57:01.949631592Z" level=info msg="loading plugin "io.containerd.snapshotter.v1.native"..." type=io.containerd.snapshotter.v1
time="2020-04-15T17:57:01.949872500Z" level=info msg="loading plugin "io.containerd.snapshotter.v1.overlayfs"..." type=io.containerd.snapshotter.v1
time="2020-04-15T17:57:01.950424047Z" level=info msg="loading plugin "io.containerd.snapshotter.v1.zfs"..." type=io.containerd.snapshotter.v1
time="2020-04-15T17:57:01.950951168Z" level=info msg="skip loading plugin "io.containerd.snapshotter.v1.zfs"..." type=io.containerd.snapshotter.v1
time="2020-04-15T17:57:01.951052585Z" level=info msg="loading plugin "io.containerd.metadata.v1.bolt"..." type=io.containerd.metadata.v1
time="2020-04-15T17:57:01.951210991Z" level=warning msg="could not use snapshotter zfs in metadata plugin" error="path /var/lib/docker/containerd/daemon/io.containerd.snapshotter.v1.zfs must be a zfs filesystem to be used with the zfs snapshotter: skip plugin"
time="2020-04-15T17:57:01.951315618Z" level=warning msg="could not use snapshotter btrfs in metadata plugin" error="path /var/lib/docker/containerd/daemon/io.containerd.snapshotter.v1.btrfs must be a btrfs filesystem to be used with the btrfs snapshotter"
time="2020-04-15T17:57:01.951365582Z" level=warning msg="could not use snapshotter aufs in metadata plugin" error="modprobe aufs failed: "ip: can't find device 'aufs'\nmodprobe: can't change directory to '/lib/modules': No such file or directory\n": exit status 1"
time="2020-04-15T17:57:01.956762009Z" level=info msg="loading plugin "io.containerd.differ.v1.walking"..." type=io.containerd.differ.v1
time="2020-04-15T17:57:01.956910038Z" level=info msg="loading plugin "io.containerd.gc.v1.scheduler"..." type=io.containerd.gc.v1
time="2020-04-15T17:57:01.957127429Z" level=info msg="loading plugin "io.containerd.service.v1.containers-service"..." type=io.containerd.service.v1
time="2020-04-15T17:57:01.957264829Z" level=info msg="loading plugin "io.containerd.service.v1.content-service"..." type=io.containerd.service.v1
time="2020-04-15T17:57:01.957342959Z" level=info msg="loading plugin "io.containerd.service.v1.diff-service"..." type=io.containerd.service.v1
time="2020-04-15T17:57:01.957444119Z" level=info msg="loading plugin "io.containerd.service.v1.images-service"..." type=io.containerd.service.v1
time="2020-04-15T17:57:01.957529606Z" level=info msg="loading plugin "io.containerd.service.v1.leases-service"..." type=io.containerd.service.v1
time="2020-04-15T17:57:01.957631974Z" level=info msg="loading plugin "io.containerd.service.v1.namespaces-service"..." type=io.containerd.service.v1
time="2020-04-15T17:57:01.957709544Z" level=info msg="loading plugin "io.containerd.service.v1.snapshots-service"..." type=io.containerd.service.v1
time="2020-04-15T17:57:01.957821858Z" level=info msg="loading plugin "io.containerd.runtime.v1.linux"..." type=io.containerd.runtime.v1
time="2020-04-15T17:57:01.958103768Z" level=info msg="loading plugin "io.containerd.runtime.v2.task"..." type=io.containerd.runtime.v2
time="2020-04-15T17:57:01.958409905Z" level=info msg="loading plugin "io.containerd.monitor.v1.cgroups"..." type=io.containerd.monitor.v1
time="2020-04-15T17:57:01.959324080Z" level=info msg="loading plugin "io.containerd.service.v1.tasks-service"..." type=io.containerd.service.v1
time="2020-04-15T17:57:01.959436139Z" level=info msg="loading plugin "io.containerd.internal.v1.restart"..." type=io.containerd.internal.v1
time="2020-04-15T17:57:01.959611248Z" level=info msg="loading plugin "io.containerd.grpc.v1.containers"..." type=io.containerd.grpc.v1
time="2020-04-15T17:57:01.959712233Z" level=info msg="loading plugin "io.containerd.grpc.v1.content"..." type=io.containerd.grpc.v1
time="2020-04-15T17:57:01.959786925Z" level=info msg="loading plugin "io.containerd.grpc.v1.diff"..." type=io.containerd.grpc.v1
time="2020-04-15T17:57:01.959881007Z" level=info msg="loading plugin "io.containerd.grpc.v1.events"..." type=io.containerd.grpc.v1
time="2020-04-15T17:57:01.959956010Z" level=info msg="loading plugin "io.containerd.grpc.v1.healthcheck"..." type=io.containerd.grpc.v1
time="2020-04-15T17:57:01.960067808Z" level=info msg="loading plugin "io.containerd.grpc.v1.images"..." type=io.containerd.grpc.v1
time="2020-04-15T17:57:01.960162509Z" level=info msg="loading plugin "io.containerd.grpc.v1.leases"..." type=io.containerd.grpc.v1
time="2020-04-15T17:57:01.960293679Z" level=info msg="loading plugin "io.containerd.grpc.v1.namespaces"..." type=io.containerd.grpc.v1
time="2020-04-15T17:57:01.960500114Z" level=info msg="loading plugin "io.containerd.internal.v1.opt"..." type=io.containerd.internal.v1
time="2020-04-15T17:57:01.960963748Z" level=info msg="loading plugin "io.containerd.grpc.v1.snapshots"..." type=io.containerd.grpc.v1
time="2020-04-15T17:57:01.961089388Z" level=info msg="loading plugin "io.containerd.grpc.v1.tasks"..." type=io.containerd.grpc.v1
time="2020-04-15T17:57:01.961196834Z" level=info msg="loading plugin "io.containerd.grpc.v1.version"..." type=io.containerd.grpc.v1
time="2020-04-15T17:57:01.961304831Z" level=info msg="loading plugin "io.containerd.grpc.v1.introspection"..." type=io.containerd.grpc.v1
time="2020-04-15T17:57:01.961654400Z" level=info msg=serving... address="/var/run/docker/containerd/containerd-debug.sock"
time="2020-04-15T17:57:01.961921360Z" level=info msg=serving... address="/var/run/docker/containerd/containerd.sock"
time="2020-04-15T17:57:01.962034717Z" level=info msg="containerd successfully booted in 0.025530s"
time="2020-04-15T17:57:01.974776267Z" level=info msg="parsed scheme: \"unix\"" module=grpc
time="2020-04-15T17:57:01.974919935Z" level=info msg="scheme \"unix\" not registered, fallback to default scheme" module=grpc
time="2020-04-15T17:57:01.975066418Z" level=info msg="ccResolverWrapper: sending update to cc: {[{unix:///var/run/docker/containerd/containerd.sock 0  <nil>}] <nil>}" module=grpc
time="2020-04-15T17:57:01.975168231Z" level=info msg="ClientConn switching balancer to \"pick_first\"" module=grpc
time="2020-04-15T17:57:01.976664640Z" level=info msg="parsed scheme: \"unix\"" module=grpc
time="2020-04-15T17:57:01.976844599Z" level=info msg="scheme \"unix\" not registered, fallback to default scheme" module=grpc
time="2020-04-15T17:57:01.977042247Z" level=info msg="ccResolverWrapper: sending update to cc: {[{unix:///var/run/docker/containerd/containerd.sock 0  <nil>}] <nil>}" module=grpc
time="2020-04-15T17:57:01.977249905Z" level=info msg="ClientConn switching balancer to \"pick_first\"" module=grpc
time="2020-04-15T17:57:02.002276709Z" level=info msg="Loading containers: start."
time="2020-04-15T17:57:02.010320736Z" level=warning msg="Running modprobe bridge br_netfilter failed with message: ip: can't find device 'bridge'\nbridge                172032  1 br_netfilter\nstp                    16384  1 bridge\nllc                    16384  2 bridge,stp\nipv6                  524288 449 nf_conntrack_ipv6,nf_nat_ipv6,bridge,[permanent]\nip: can't find device 'br_netfilter'\nbr_netfilter           24576  0 \nbridge                172032  1 br_netfilter\nmodprobe: can't change directory to '/lib/modules': No such file or directory\n, error: exit status 1"
time="2020-04-15T17:57:02.223181931Z" level=info msg="Default bridge (docker0) is assigned with an IP address 172.17.0.0/16. Daemon option --bip can be used to set a preferred IP address"
time="2020-04-15T17:57:02.410953878Z" level=info msg="Loading containers: done."
time="2020-04-15T17:57:02.455350319Z" level=info msg="Docker daemon" commit=369ce74a3c graphdriver(s)=overlay2 version=19.03.6
time="2020-04-15T17:57:02.455540195Z" level=info msg="Daemon has completed initialization"
time="2020-04-15T17:57:02.481630177Z" level=info msg="API listen on [::]:2376"
time="2020-04-15T17:57:02.481781411Z" level=info msg="API listen on /var/run/docker.sock"
```

</details>

**Comment by mumoshu** [comment]
@chenrui333 The runner pod is taught to restart the container on failure https://github.com/summerwind/actions-runner-controller/blob/5714459c24e7cad727eeefe2d188aedcadb47c39/controllers/runner_controller.go#L275 and the runner is given `--once` to stop on successful run.

So I suppose this is due to the `runner` container failing, not the dind container.

Would you mind sharing full logs from failed `runner` containers?

 I guess `kubectl logs --previous -c runner` would work.

**Comment by chenrui333** [comment]
Actually, after bumping the runner to latest version today (v2.169.1), I have the runner working for my favor now (even though the runner still recreating the containers from time to time).

**Comment by chenrui333** [comment]
Since I have already bumped the fleet to the latest version, `restart` is no longer issue for me.

**Comment by chenrui333** [comment]
The container recreation would be something like this (it is re-created really fast though):

![image](https://user-images.githubusercontent.com/1580956/79406495-47212380-7f65-11ea-8b0e-5be758ffd078.png)

![image](https://user-images.githubusercontent.com/1580956/79406476-35d81700-7f65-11ea-99d8-c35a841b2d29.png)

**Comment by mumoshu** [comment]
@chenrui333 Thanks for reporting!

I believe the pod recreation you're seeing currently is an expected behavior.

We do configure the runner agent to stop after each workflow run, so that the controller can recreate it and bring up an another, clean runner pod, which is important for reproducible builds.

**Comment by mumoshu** [comment]
@chenrui333 Regarding the original issue, I got to think that it's due to update failure. Reading the runner code, I see it exits with a non-zero code(3) as I've explained in https://github.com/summerwind/actions-runner-controller/issues/33#issuecomment-614585687.

That would result in the runner container(not dind container or the pod) to be stopped and restarted.

**Comment by chenrui333** [comment]
yeah, I feel the same way too, we need to do something to stop it self running. 

But the container does get re-created pretty often though. :)

**Comment by mumoshu** [comment]
> But the container does get re-created pretty often though. :)

I'd say it's an expected behaviour as long as it's getting recreated only after each workflow run.

Or does it get recreated without any workflow run nor any error?

**Comment by mumoshu** [comment]
Runner updates seem to be fine these days and I believe we learned a lot about how we should deal with it since then.

Regarding the ability to not restart the runner on each job run at all, we have another issue #457.

Closing this as resolved! Thanks for reporting.
